### ISSUE TITLE
Backward pass convs have two reduces

### ISSUE THREAD

**Issue by geohot** [tinygrad/tinygrad#3572]
See in beautiful_mnist

**Comment by geohot** [comment]
![Screenshot 2024-03-02 at 5 00 44 AM](https://github.com/tinygrad/tinygrad/assets/72895/79572994-26a5-44cf-98b4-1ba71e1b3545)

This is the pattern

**Comment by davidjanoskyrepo** [comment]
I'm not sure how to view the graphs but I would think the double reduce comes from the double max_pool2d. Modifying the model to this:

def __init__(self, use_new:bool=False) -> None:
    input_dim = 32
    expand_dim = 64
    lin_size = 9
    self.layers: List[Callable[[Tensor], Tensor]] = [
      nn.Conv2d(1, input_dim, 5), Tensor.relu,
      # Approximate the effect of the maxpooling's reduction in size by
      # increasing the stride of the convolution
      nn.Conv2d(input_dim, input_dim, 5, stride=1 if not use_new else 2), Tensor.relu,
      nn.BatchNorm2d(input_dim),
      # Remove first maxpooling, which increases the parameter size but only does one reduce
      Tensor.max_pool2d if not use_new else lambda x: x,  # NOTE: this is a hack to make the JIT work
      nn.Conv2d(input_dim, expand_dim, 3), Tensor.relu,
      nn.Conv2d(expand_dim, expand_dim, 3), Tensor.relu,
      nn.BatchNorm2d(expand_dim), Tensor.max_pool2d,
      lambda x: x.flatten(1), nn.Linear(expand_dim*lin_size, 10)]

Results in a speedup without measurable loss and the same number of parameters:

Training old model
Number of parameters: 88044
loss:   0.08 test_accuracy: 98.21%: 100%|██████████████| 70/70 [00:07<00:00,  8.90it/s] 
Training old model
Number of parameters: 88044
loss:   0.06 test_accuracy: 97.79%: 100%|██████████████| 70/70 [00:06<00:00, 10.43it/s] 
Training new model
Number of parameters: 88044
loss:   0.07 test_accuracy: 97.90%: 100%|██████████████| 70/70 [00:04<00:00, 15.64it/s] 
Training new model
Number of parameters: 88044
loss:   0.08 test_accuracy: 98.02%: 100%|██████████████| 70/70 [00:04<00:00, 16.60it/s]

**Comment by Itssshikhar** [comment]
Hi! just wanted to know that if backward-pass needs both the reduceOps replaced or removed for perf?

**Comment by iglesias** [comment]
> Hi! just wanted to know that if backward-pass needs both the reduceOps replaced or removed for perf?

I'd guess it is about simplifying the two reduces into one.

**Comment by Itssshikhar** [comment]
Based on the reduce operations(SUM) in beautiful_mnist, I think some changes would be needed to make in gradients update Adam optimiser itself.
`self.m[i].assign(self.b1 * self.m[i] + (1.0 - self.b1) * t.grad)
self.v[i].assign(self.b2 * self.v[i] + (1.0 - self.b2) * (t.grad * t.grad))`.
Would like to know if I can get ahead with this as a proposed change?

**Comment by Qazalin** [comment]
https://github.com/tinygrad/tinygrad/pull/4441
